### PR TITLE
Elide memcpy where not needed

### PR DESCRIPTION
**Pull request by thorstenhater**
TL;DR: Simple change for ~2% gain on GPU in a typical workload.

This removes resetting Xi, Xo, and eX when these have not been modified by any mechanism.
In most simulations either none or only one of these is written. Likely this is only a minor gain 
on the CPU backend, but on the GPU it's a different story.

By inspecting a recent profiling run I found of the overall runtime 
- 13% in memcpy D2H/H2D
- 3% in D2D

for 3 ion species (`K`, `Ca`, and `Na`). Out of those, only `cai` and `ena` where written.
This PR therefore elides 7 out of 9 D2D memcpys related to ion reset in this case.

### CODE REVIEW

**Comment by thorstenhater** [comment]
bors try

File: arbor/backends/gpu/shared_state.hpp [hunk 1]
```diff
@@ -27,8 +27,11 @@ namespace gpu {
  *     Xi_     cai              internal calcium concentration
  *     Xo_     cao              external calcium concentration
  */
+ struct ARB_ARBOR_API ion_state {
+    bool write_eX;          // is eX written?
+    bool write_Xo;          // is Xo written?
+    bool write_Xi;          // is Xi written?
```
Comment (bcumming) [inline arbor/backends/gpu/shared_state.hpp:33]: Rename to have tralining underscore, e.g. `write_eX_`, to be in line with naming scheme of other member variables.

File: arbor/backends/multicore/shared_state.hpp [hunk 2]
```diff
@@ -38,10 +38,13 @@ namespace multicore {
  *     Xi_     cai              internal calcium concentration
  *     Xo_     cao              external calcium concentration
  */
-
 struct ARB_ARBOR_API ion_state {
     unsigned alignment = 1; // Alignment and padding multiple.
 
+    bool write_eX;          // is eX written?
+    bool write_Xo;          // is Xo written?
+    bool write_Xi;          // is Xi written?
```
Comment (bcumming) [inline arbor/backends/multicore/shared_state.hpp:46]: Add trailing underscore.

File: arbor/fvm_layout.cpp [hunk 3]
```diff
@@ -1029,6 +1036,15 @@ fvm_mechanism_data fvm_build_mechanism_data(
         // If synapse uses an ion, add to ion support.
         update_ion_support(info, config.cv);
 
+        for (const auto& [ion, dep]: info.ions) {
+            if (dep.write_concentration_int) {
+                write_xi.insert(ion);
+            }
+            if (dep.write_concentration_ext) {
+                write_xi.insert(ion);
```
Comment (bcumming) [inline arbor/fvm_layout.cpp:1044]: Should this be `write_xo.insert(ion)`, i.e. not `xi`?

File: arbor/fvm_layout.cpp [hunk 4]
```diff
@@ -1091,6 +1107,16 @@ fvm_mechanism_data fvm_build_mechanism_data(
             }
             lid_junction_desc.insert({pm.lid, std::move(per_lid)});
         }
+
+        for (const auto& [ion, dep]: info.ions) {
+            if (dep.write_concentration_int) {
+                write_xi.insert(ion);
+            }
+            if (dep.write_concentration_ext) {
+                write_xi.insert(ion);
```
Comment (bcumming) [inline arbor/fvm_layout.cpp:1116]: Should this be write_xo.insert(ion), i.e. not xi?

File: arbor/fvm_layout.cpp [hunk 5]
```diff
@@ -1269,6 +1295,8 @@ fvm_mechanism_data fvm_build_mechanism_data(
                 throw cable_cell_error("revpot mechanism for ion "+ion+" does not write this reversal potential");
             }
 
+            M.ions[ion].revpot_written = true;
```
Comment (bcumming) [inline arbor/fvm_layout.cpp:1298]: This is always set to true from my reading if the code. Is that meant to be?

**Review by bcumming — CHANGES_REQUESTED** [review]
Looks good.

A few comments and possible fixes.

File: arbor/fvm_layout.cpp (hunk 5 above)
Comment (thorstenhater) [reply arbor/fvm_layout.cpp:1298]: Yes, that's why this writes directly to `M.ions` and does not go through a intermediary. Or did I misread your comment?

**Comment by thorstenhater** [comment]
bors try

**Comment by thorstenhater** [comment]
bors try

**Comment by thorstenhater** [comment]
bors try

**Comment by thorstenhater** [comment]
bors try

**Comment by bcumming** [comment]
bors try

**Review by bcumming — APPROVED** [review]

**Comment by brenthuisman** [comment]
bors try

**Pull request merged**